Swap `validator` for `garde`

mateiidavid opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review the requirements below.

Bug fixes and new features should include tests.

Contributors guide: https://github.com/kube-rs/kube-rs/blob/master/CONTRIBUTING.md
-->

## Motivation

The validator crate used in the derive docs and in the crd example is no longer actively maintained. It has been superseded by garde which offers similar functionality.

Fixes #1197 (check issue for additional context)
<!--
Explain the context and why you're making that change. What is the problem you're trying to solve?
If a new feature is being added, describe the intended use case that feature fulfills.
-->

## Solution

This change replaces validator with garde in dev dependencies (and in the crd_api example dependency). Although garde has the same derive macro, its attribute differs from validator, and it additionally requires fields to be explictly skipped from validation, and does not offer a built-in way to validate required fields.

Changes have been made to the kube_derive crate's docs to both replace existing mentions of validator, and inform on the caveats mentioned above.

<!--
Summarize the solution and provide any necessary context needed to understand the code change.
-->

clux commented on kube/src/lib.rs:
```diff
@@ -62,9 +62,12 @@
 //! pub struct FooSpec {
+//!     #[garde(skip)]
 //!     info: String,
 //!     #[validate(length(min = 3))]
+//!     #[garde(length(min = 3))]
```
Slightly sad about this as these attributes are identical, but it does technically provide us with a new way to separate client-side and server-side validation (if that is useful), even though we were primarily using it as a way to hook into schemars.

Not sure if there's anything schemars or garde can do about this. I saw no mention of [schemars in garde](https://github.com/jprochazk/garde/search?q=schemars), and similarly no mention of the [garde in schemars](https://github.com/GREsau/schemars/search?q=garde).

There is [this doc](https://github.com/GREsau/schemars/blob/73715c10e0007c766969732a1ff53f358c807780/docs/1.1-attributes.md#supported-validator-attributes) that shows that we can use the `#[schemars(length(min = 3))]` instead to highlight that this is a schema-only feature which might be worth doing to avoid confusion over the fork.

mateiidavid replied:
Yeah, it is a bit unfortunate. It's nice that we get a separation, but it was also really cool to have just one attribute that can be used for validation on both sides. Maybe at some point something can be PR'd upstream to support `#[garde]` in schemars.

Great find with the doc! I worked completely off the git history and was under the impression they only support `#[validate]`. Will make the changes 👍🏻

clux commented on kube/Cargo.toml:
```diff
@@ -50,8 +50,8 @@ default-features = false
 [dev-dependencies]
 tokio = { version = "1.14.0", features = ["full"] }
 futures = "0.3.17"
+garde = { version = "0.11.2", default-features = false, features = ["derive"] }
```
if this only does client-side validation now, we should probably just take it out and leave `garde` as an example nicety as it currently is in `crd_api.rs`.

the elaborating doc comments in kube-derive are good, but we can probably remove references to `garde` in the actual doc examples for one less crate to build.

clux commented on kube-derive/src/lib.rs:
```diff
@@ -230,14 +234,18 @@ mod custom_resource;
 /// ## Validation Caveats
-/// The supported **`#[validate]` attrs also exist as `#[schemars]` attrs** so you can use those directly if you do not require the validation to run client-side (in your code).
-/// Otherwise, you should `#[derive(Validate)]` on your struct to have both server-side (kubernetes) and client-side validation.
+/// A number of `#[validate]` attrs are supported by `schemars` and can be used directly on your struct if you do not require validation to run client-side (in your code).
+/// Otherwise, to have both server-side (kubernetes) and client-side validation, you should `#[derive(Validate)]` on your struct and include additional `#[garde]` attrs that mirror the `#[validate]` ones.
+/// Two different sets of attrs are needed because `schemar`'s validation attrs are based on a now unmaintained [`validator`](https://github.com/Keats/validator/issues/201) crate.
```
Let us highlight that there are two avenues now that solve two different problems instead.

```suggestion
/// There are two main ways of doing validation; **server-side** (embedding validation attributes into the schema for the apiserver to respect), and **client-side** (provides validate() methods in your code).
///
/// Client side validation of structs can be achieved by hooking up `#[garde]` attributes in your struct and is a replacement of the now unmaintained [`validator`](https://github.com/Keats/validator/issues/201) crate.
/// Server-side validation require mutation of your generated schema, and can in the basic cases be achieved through the use of `schemars`'s [validation attributes](https://graham.cool/schemars/deriving/attributes/#supported-validator-attributes).
/// For complete control, [parts of the schema can be overridden](https://github.com/kube-rs/kube/blob/e01187e13ba364ccecec452e023316a62fb13e04/examples/crd_derive.rs#L37-L38) to support more advanced [Kubernetes specific validation rules](https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/)
```

possibly needs a rework with the parent header though

mateiidavid replied:
Nice 👍🏻 I applied the suggestion as is and changed the parent header to `## Schema Validation`. Added a nested header for caveats.

clux reviewed (commented):
Thanks for taking the time to dig into this. I think that given how clearly this ends up separating client-side and server-side validation, i think we should take garde out of our dev-dependencies and leave it as an example trick, but the docs are good. Left one bigger reword.

clux commented on examples/crd_api.rs:
```diff
@@ -23,8 +23,11 @@ use kube::{
 #[kube(printcolumn = r#"{"name":"Team", "jsonPath": ".spec.metadata.team", "type": "string"}"#)]
 pub struct FooSpec {
     #[validate(length(min = 3))]
```
let's change this (and similar ones) to `#[schemars(length(min = 3))]`

mateiidavid replied:
ah, nice. I should've dug into schemars' docs a bit more, this makes more sense.

mateiidavid commented:
@clux thanks for the review! I made a few changes:

* removed garde from dev deps
* replaced `#[validate]` with `#[schemars]` in both comments and code
* re-worked the docs a bit

Lmk how it looks, happy to finesse this more :)

clux reviewed (approved):
Looks great now. Thank you!
